src/components/product-progress/product-progress.service.ts

```diff
@@ -16,34 +11,44 @@ export class ProductProgressService {
     reportId: ID,
     session: Session
   ): Promise<ProductProgress[]> {
-    throw new NotImplementedException().with(reportId, session);
+    const productIds = await this.repo.readAllProgressReportsByReport(reportId);
+    return await Promise.all(
+      productIds.map(
+        async (productId) => await this.readOne(reportId, productId, session)
+      )
+    );
   }
 
   async readAllByProduct(
     productId: ID,
     session: Session
   ): Promise<ProductProgress[]> {
-    throw new NotImplementedException().with(productId, session);
+    const reportIds = await this.repo.readAllProgressReportsByProduct(
+      productId
+    );
+    return await Promise.all(
+      reportIds.map(
+        async (reportId) => await this.readOne(reportId, productId, session)
+      )
+    );
   }
 
   async readOne(
     reportId: ID,
     productId: ID,
-    session: Session
+    _session: Session
   ): Promise<ProductProgress> {
-    throw new NotImplementedException().with(reportId, productId, session);
-  }
+    const steps = await this.repo.readSteps(productId, reportId);
 
-  async update(input: ProductProgressInput, session: Session) {
-    // TODO ensure all input.steps[].step are set on the product.
-    // TODO ensure input.reportId is valid
-
-    // Convert steps list to map. Use if useful, otherwise remove.
-    const steps = mapFromList(input.steps, (pair) => [
-      pair.step,
-      pair.percentDone,
-    ]);
+    return {
+      productId,
+      reportId,
+      steps,
+    };
+  }
 
-    throw new NotImplementedException().with(input, session, steps);
+  async update(input: ProductProgressInput) {
+    await this.repo.removeOldProgress(input.productId, input.reportId);
+    await this.repo.create(input);
   }
```

Comment:
Ah yeah we don't want to delete & recreate the `ProductProgress` node on update. I'm not sure if I stated this but I was thinking that the steps input array could be partial - only changing / setting values for the steps given. Or maybe you coded the UI to send all the step values at once?

Maybe we do a merge and on create set the ID and createdAt?